buildSrc/src/main/groovy/com/google/firebase/gradle/GenerateMeasurementsTask.groovy

```diff
@@ -0,0 +1,182 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+package com.google.firebase.gradle
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.Task
+import org.gradle.api.tasks.InputFile
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.TaskAction
+
+/**
+ * Generates size measurements after building the test apps and outputs them as a text-format
+ * protocol buffer report.
+ *
+ * <p>This task requires two properties, an SDK map, as input, and the report, as output. The map is
+ * used to convert project names and build variants into the SDK identifiers used by the database.
+ * The report path is where the output should be stored. Additionally, a project property, {@code
+ * pull_request} is also required. This value will be placed in the report.
+ */
+public class GenerateMeasurementsTask extends DefaultTask {
+
+    /**
+     * The file storing the SDK map.
+     *
+     * <p>This may be any type recognized by Gradle as a file. The format of the file's contents is
+     * headerless CSVwith a colon as a delimiter: projectName_buildVariant:sdkId. The first column
+     * contains both the project name and build variant separated by an hyphen, {@code
+     * firestore-aggressive}, for example.
+     */
+    @InputFile
+    File sdkMap
+
+    /**
+     * The file for storing the report.
+     *
+     * <p>This may be any type recognized by Gradle as a file. The contents, if any, will be
+     * overwritten by the new report.
+     */
+    @OutputFile
+    File report
+
+    @Override
+    Task configure(Closure closure) {
+        project.android.variantFilter {
+            if (it.buildType.name != "aggressive") {
+                it.ignore = true;
+            }
+        }
+
+        outputs.upToDateWhen { false }
+        dependsOn "assemble"
+        return super.configure(closure)
+    }
+
+    @TaskAction
+    def generate() {
+        def sdks = createSdkMap()
+        def sizes = calculateSizes(sdks, project.android.applicationVariants)
+        def report = createReport(sizes)
+
+        project.file(this.report).withWriter {
+            it.write(report)
+        }
+    }
+
+    private def calculateSizes(sdks, variants) {
+      def sizes = [:]
+
+      variants.each { variant ->
+        def name = "${variant.flavorName}-${variant.buildType.name}"
+        def apks = variant.outputs.findAll { it.outputFile.name.endsWith(".apk") }
+        if (apks.size() > 1) {
+            throw new IllegalStateException("${name} produced more than one APK")
+        }
```

Comment:
Do we fail silently if < 1?

Reply:
We skip that variant if there's no APK. I'll add a comment better clarifying the behavior.

I don't think there's a reason for the uploader to fail if a previous build step failed to produce an APK.